cri/v1alpha1/cri_utils.go

```diff
@@ -940,3 +941,137 @@ func blkrdev(device string) (uint64, error) {
+		KernelMemory:          apitypesResources.KernelMemory,
+		MemoryReservation:     apitypesResources.MemoryReservation,
+		MemorySwappiness:      memorySwappiness,
+		Ulimits:               parseUlimitFromPouch(apitypesResources.Ulimits),
+		DiskQuota:             diskQuota,
+	}
+}
+
+// parseWeightDeviceFromAPI parse WeightDevice from runtime.WeightDevice to apitypes.WeightDevice
+func parseWeightDeviceFromAPI(runtimeWeightDevices []*runtime.WeightDevice) (weightDevices []*apitypes.WeightDevice) {
+	for _, v := range runtimeWeightDevices {
+		weightDevices = append(weightDevices, &apitypes.WeightDevice{
+			Path:   v.GetPath(),
+			Weight: uint16(v.GetWeight()),
+		})
+	}
+	return
+}
+
+// parseWeightDeviceFromPouch parse WeightDevice from apitypes.WeightDevice to runtime.WeightDevice
+func parseWeightDeviceFromPouch(apitypesWeightDevices []*apitypes.WeightDevice) (weightDevices []*runtime.WeightDevice) {
+	for _, v := range apitypesWeightDevices {
+		weightDevices = append(weightDevices, &runtime.WeightDevice{
+			Path:   v.Path,
+			Weight: uint32(v.Weight),
+		})
+	}
+	return
+}
+
+// parseThrottleDeviceFromAPI parse ThrottleDevice from runtime.ThrottleDevice to apitypes.ThrottleDevice
+func parseThrottleDeviceFromAPI(runtimeThrottleDevices []*runtime.ThrottleDevice) (throttleDevices []*apitypes.ThrottleDevice) {
+	for _, v := range runtimeThrottleDevices {
+		throttleDevices = append(throttleDevices, &apitypes.ThrottleDevice{
+			Path: v.GetPath(),
+			Rate: v.GetRate(),
+		})
+	}
+	return
+}
+
+// parseThrottleDeviceFromPouch parse ThrottleDevice from apitypes.ThrottleDevice to runtime.ThrottleDevice
+func parseThrottleDeviceFromPouch(apitypesThrottleDevices []*apitypes.ThrottleDevice) (throttleDevices []*runtime.ThrottleDevice) {
+	for _, v := range apitypesThrottleDevices {
+		throttleDevices = append(throttleDevices, &runtime.ThrottleDevice{
+			Path: v.Path,
+			Rate: v.Rate,
+		})
+	}
+	return
+}
+
+// parseUlimitFromAPI parse Ulimit from runtime.Ulimit to apitypes.Ulimit
+func parseUlimitFromAPI(runtimeUlimits []*runtime.Ulimit) (ulimits []*apitypes.Ulimit) {
+	for _, v := range runtimeUlimits {
+		ulimits = append(ulimits, &apitypes.Ulimit{
+			Hard: v.GetHard(),
+			Name: v.GetName(),
+			Soft: v.GetSoft(),
+		})
+	}
+	return
+}
+
+// parseUlimitFromPouch parse Ulimit from apitypes.Ulimit to runtime.Ulimit
+func parseUlimitFromPouch(apitypesUlimits []*apitypes.Ulimit) (ulimits []*runtime.Ulimit) {
+	for _, v := range apitypesUlimits {
+		ulimits = append(ulimits, &runtime.Ulimit{
+			Hard: v.Hard,
+			Name: v.Name,
+			Soft: v.Soft,
+		})
+	}
+	return
+}
+
+// parseVolumesFromPouch parse Volumes from map[string]interface{} to map[string]*runtime.Volume
+func parseVolumesFromPouch(containerVolumes map[string]interface{}) map[string]*runtime.Volume {
+	volumes := make(map[string]*runtime.Volume)
+	for k, v := range containerVolumes {
+		volumes[k] = v.(*runtime.Volume)
```

Comment:
always `(*runtime.Volume)` ? nil or other type?

Reply:
Done. thanks for your comment.